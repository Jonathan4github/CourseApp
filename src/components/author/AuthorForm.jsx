import React from 'react';
import PropTypes from 'prop-types';

import TextInput from '../common/TextInput';


const propTypes = {
  author: PropTypes.shape().isRequired,
  handleOnSave: PropTypes.func.isRequired,
  handleOnChange: PropTypes.func.isRequired,
  handleOnFocus: PropTypes.func.isRequired,
  isSaving: PropTypes.bool.isRequired,
  errors: PropTypes.shape()
};

const defaultProps = {
  errors: {}
};

const AuthorForm = ({
  author, handleOnSave, handleOnChange, handleOnFocus, isSaving, errors
}) => (
  <form>
    <TextInput
      name="firstName"
      label="First Name"
      value={author.firstName}
      onChange={handleOnChange}
      onFocus={handleOnFocus}
      placeholder="eg. John"
      error={errors.firstName} />
    <TextInput
      name="lastName"
      label="Last Name"
      value={author.lastName}
      onChange={handleOnChange}
      onFocus={handleOnFocus}
      placeholder="eg. Doe"
      error={errors.lastName} />
    <input
     type="submit"
     className="btn btn-primary"
     disabled={isSaving}
     value={isSaving ? 'Saving...' : 'Save'}
     onClick={handleOnSave} />
  </form>
);
AuthorForm.propTypes = propTypes;
AuthorForm.defaultProps = defaultProps;

export default AuthorForm;
