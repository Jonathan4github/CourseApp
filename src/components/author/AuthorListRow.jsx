import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';


const propTypes = {
  author: PropTypes.shape().isRequired,
  handleDelete: PropTypes.func.isRequired
};

const AuthorListRow = ({ author, handleDelete }) => (
  <tr>
    <td><Link to={`/authors/${author.id}`}>{author.id}</Link></td>
    <td>{author.firstName}</td>
    <td>{author.lastName}</td>
    <td>
      <button
        type="button"
        onClick={() => handleDelete(author.id)}
        className="btn btn-outline-danger btn-sm">
        Delete
      </button>
    </td>
  </tr>
);

AuthorListRow.propTypes = propTypes;

export default AuthorListRow;
