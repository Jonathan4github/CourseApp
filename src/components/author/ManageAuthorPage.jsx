import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { Prompt } from 'react-router-dom';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import Loader from 'react-md-spinner';

import AuthorForm from './AuthorForm';
import * as authorActions from '../../actions/creators/authorActions';

import { validateFormData } from '../../helpers/validations';


const propTypes = {
  actions: PropTypes.shape({
    loadSingleAuthor: PropTypes.func.isRequired,
    saveAuthor: PropTypes.func.isRequired,
    loadAuthors: PropTypes.func.isRequired
  }).isRequired,
  author: PropTypes.shape({
    selectedAuthor: PropTypes.shape().isRequired,
    isFetching: PropTypes.bool.isRequired,
    isSaving: PropTypes.bool.isRequired
  }).isRequired,
  allAuthors: PropTypes.array.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape().isRequired
  }).isRequired,
  history: PropTypes.shape({ push: PropTypes.func.isRequired }).isRequired
};

class ManageAuthorPage extends Component {
  initialFormState = {
    author: {
      id: '',
      firstName: '',
      lastName: ''
    },
    errors: {},
    isBlocking: false
  };

  state = { ...this.initialFormState };

  componentDidMount() {
    const { match, actions } = this.props;

    if (match.params.id !== 'add') {
      actions.loadSingleAuthor(match.params.id);
    }
  }

  componentWillReceiveProps(nextProps) {
    const { match, author: { selectedAuthor } } = this.props;
    const { author } = nextProps;
    if (match.params.id !== 'add') {
      if (author.selectedAuthor !== selectedAuthor) {
        this.setState(prevState => ({
          ...prevState,
          author: nextProps.author.selectedAuthor
        }));
      }
    } else {
      this.setState(prevState => ({
        ...prevState
      }));
    }
  }

  handleOnChange = (event) => {
    const { name, value } = event.target;
    this.setState(prevState => ({
      ...prevState,
      author: { ...prevState.author, [name]: value },
      isBlocking: true
    }));
  };

  isValid = (author) => {
    const { errors, isValid } = validateFormData(author);
    if (!isValid) {
      this.setState(() => ({
        errors
      }));
    }
    return isValid;
  }

  handleOnFocus = (event) => {
    event.persist();
    this.setState(prevState => ({
      errors: { ...prevState.errors, [event.target.name]: '' }
    }));
  }

  handleOnSave = (event) => {
    const { actions, history } = this.props;
    const { author } = this.state;

    event.preventDefault();

    if (this.isValid(author)) {
      actions.saveAuthor(author)
        .then(() => {
          this.setState(prevState => ({ isBlocking: !prevState }), () => {
            history.push('/authors');
          });
        });
    }
  };

  render() {
    const { author, errors, isBlocking } = this.state;
    const { allAuthors, author: { isFetching, isSaving } } = this.props;
    return (
      <Fragment>
        <Prompt
          when={isBlocking}
          message={
            location => `Are you sure you want to go to ${location.pathname}`
          }
        />
        <div className="top-container">
          <h1>Manage Author</h1>
        </div>
        {isFetching
          ? (
            <div className="loader-container">
              <Loader size="30" className="loader-component" />
            </div>
          )
          : (
            <AuthorForm
              allAuthors={allAuthors}
              author={author}
              handleOnChange={this.handleOnChange}
              handleOnSave={this.handleOnSave}
              handleOnFocus={this.handleOnFocus}
              isSaving={isSaving}
              errors={errors} />
          )
        }
      </Fragment>
    );
  }
}

const mapStateToProps = state => ({
  author: state.authors,
  allAuthors: state.authors.allAuthors
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(authorActions, dispatch)
});

ManageAuthorPage.propTypes = propTypes;

export default connect(mapStateToProps, mapDispatchToProps)(ManageAuthorPage);
