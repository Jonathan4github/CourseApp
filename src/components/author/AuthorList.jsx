import React from 'react';
import PropTypes from 'prop-types';
import AuthorListRow from './AuthorListRow';


const propTypes = {
  authors: PropTypes.array.isRequired,
  handleDelete: PropTypes.func.isRequired
};

const AuthorList = ({ authors, handleDelete }) => (
  <table className="table table-hover">
    <thead>
      <tr>
        <th>ID</th>
        <th>First Name</th>
        <th>Last Name</th>
        <th>Action</th>
      </tr>
    </thead>
    <tbody>
      {authors
        .map(author => (
          <AuthorListRow
            key={author.id}
            author={author}
            handleDelete={handleDelete} />
        ))
      }
    </tbody>
  </table>
);

AuthorList.propTypes = propTypes;

export default AuthorList;
