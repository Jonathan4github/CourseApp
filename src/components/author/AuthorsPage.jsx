import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import Loader from 'react-md-spinner';
import _ from 'lodash';

import AuthorList from './AuthorList';
import NoItems from '../common/NoItems';
import TablePagination from '../common/TablePagination';
import * as authorActions from '../../actions/creators/authorActions';
import { AUTHORS_PAGE_CHANGE as actionType } from '../../actions/constants/actionTypes';
import pagination from '../../helpers/pagination';

const propTypes = {
  authors: PropTypes.shape({
    allAuthors: PropTypes.array.isRequired,
    isFetching: PropTypes.bool.isRequired,
    pageCount: PropTypes.number.isRequired,
    pageSize: PropTypes.number.isRequired,
    currentPage: PropTypes.number.isRequired
  }).isRequired,
  actions: PropTypes.shape({
    deleteAuthor: PropTypes.func.isRequired,
    pageChange: PropTypes.func.isRequired
  }).isRequired,
  history: PropTypes.shape({ push: PropTypes.func.isRequired }).isRequired
};

class AuthorsPage extends Component {
 actions = this.props.actions;

 componentWillUpdate({ authors: { pageSize, pageCount, currentPage } }) {
   if (pageSize === 0 && currentPage > 1 && pageCount >= 1) {
     this.actions.pageChange(currentPage - 1);
   }
 }

 componentWillUnmount() {
   this.actions.pageChange(1, actionType);
 }

  showAddAuthorPage = () => {
    const { history } = this.props;
    history.push('/authors/add');
  };

  handleOnAuthorDelete = (authorId) => {
    this.actions.deleteAuthor(authorId);
  };

  handleOnPageChange = ({ selected }) => {
    const pageToLoad = selected + 1;
    this.actions.pageChange(pageToLoad, actionType);
  }

  renderAuthorsTable = (paginatedAuthors) => {
    if (paginatedAuthors.length > 0) {
      return (
        <AuthorList
          authors={paginatedAuthors}
          handleDelete={this.handleOnAuthorDelete} />
      );
    }
    return <NoItems displayText="There are no registered authors" />;
  }

  render() {
    const {
      authors: {
        allAuthors,
        isFetching,
        currentPage,
        pageCount
      }
    } = this.props;
    return (
      <div>
        <div className="top-container">
          <h2>Authors</h2>
          <button
            type="button"
            className="btn btn-outline-primary add-course-button btn-sm"
            onClick={this.showAddAuthorPage}>
            Add Author
          </button>
        </div>
        {isFetching
          ? (
            <div className="loader-container">
              <Loader size="30" className="loader-component" />
            </div>
          )
          : this.renderAuthorsTable(allAuthors)
        }
        {allAuthors.length > 0 && (
          <TablePagination
            pageCount={pageCount}
            currentPage={currentPage}
            handlePageChange={this.handleOnPageChange}
          />
        )}
      </div>
    );
  }
}

const mapStateToProps = ({ authors }) => {
  const sortedAuthors = _.sortBy(
    [...authors.allAuthors], author => author.firstName.toLowerCase()
  );

  const data = pagination(sortedAuthors, authors.currentPage);

  return ({
    authors: {
      ...authors,
      allAuthors: data.paginatedItems,
      pageCount: data.totalPageCount,
      pageSize: data.pageItemsSize
    }
  });
};

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(authorActions, dispatch)
});

AuthorsPage.propTypes = propTypes;

export default connect(mapStateToProps, mapDispatchToProps)(AuthorsPage);
