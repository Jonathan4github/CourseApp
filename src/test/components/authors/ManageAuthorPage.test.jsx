/* eslint-disable */
import { ManageAuthorPage } from '../../../components/author/ManageAuthorPage';
import authorsMock from '../../__mocks__/authorMock';


describe('<ManageAuthorPage /> component', () => {
  const setup = () => {
    const props = {
      actions: {
        loadSingleAuthor: jest.fn(),
        saveAuthor: jest.fn(),
        loadAuthors: jest.fn()
      },
      author: {
        selectedAuthor: {},
        isFetching: false,
        isSaving: false
      },
      allAuthors: authorsMock.authors,
      match: {
        params: {
          id: 'add'
        }
      },
      history: {
        push: jest.fn()
      }
    };

    const state = {
      errors: {},
      isBlocking: false
    };

    return {
      pressentationWrapper: shallow(<ManageAuthorPage {...props} />),
      props,
      state
    };
  };

  const event = {
    target: {
      name: 'firstName',
      value: 'some name'
    },
    preventDefault: jest.fn(),
    persist: jest.fn()
  };

  it('render component without crashing', () => {
    const { pressentationWrapper } = setup();

    expect(toJson(pressentationWrapper)).toMatchSnapshot();
  });

  it('invokes `componentDidMount` when mounted', () => {
    jest.spyOn(ManageAuthorPage.prototype, 'componentDidMount');
    setup();

    expect(ManageAuthorPage.prototype.componentDidMount).toHaveBeenCalled();
    ManageAuthorPage.prototype.componentDidMount.mockRestore();
  });

  it('renders a <Prompt /> child component', () => {
    const { pressentationWrapper } = setup();
    expect(pressentationWrapper.find('Prompt').length).toBe(1);
  });

  it('should show <Prompt /> component if value of its child prop(when) is true', () => {
    const { pressentationWrapper, state } = setup();

    pressentationWrapper.setState({
      ...state, isBlocking: true
    });

    expect(pressentationWrapper.find('Prompt').props().when).toEqual(true);
  });

  it('invokes handleOnChange method', () => {
    const { pressentationWrapper } = setup();
    const handleOnChangeSpy = jest
      .spyOn(pressentationWrapper.instance(), 'handleOnChange');
    pressentationWrapper.instance().handleOnChange(event);

    expect(handleOnChangeSpy).toHaveBeenCalled();
  });

  it('invokes handleOnSave method', () => {
    const { pressentationWrapper } = setup();
    const handleOnSaveSpy = jest.spyOn(pressentationWrapper.instance(), 'handleOnSave');
    pressentationWrapper.instance().handleOnSave(event);
    expect(handleOnSaveSpy).toHaveBeenCalled();
  });

  it('invokes handleOnFocus method', () => {
    const { pressentationWrapper } = setup();
    const handleOnFocusSpy = jest.spyOn(pressentationWrapper.instance(), 'handleOnFocus');
    pressentationWrapper.instance().handleOnFocus(event);
    expect(handleOnFocusSpy).toHaveBeenCalled();
  });

  it('shows <Loader /> component when `isFetching` is true', () => {
    const { pressentationWrapper, props } = setup();
    pressentationWrapper.setProps({
      ...props,
      author: { ...props.author, isFetching: true }
    });

    expect(pressentationWrapper.find('.loader-container')).toHaveLength(1);
  });

  it('shows <NotFoundPage /> component when `authorId` does not exist', () => {
    const { pressentationWrapper, props } = setup();
    pressentationWrapper.setProps({
      ...props,
      author: { ...props.author, selectedAuthor: undefined }
    });

    expect(pressentationWrapper.find('NotFoundPage')).toHaveLength(1);
  });
});
