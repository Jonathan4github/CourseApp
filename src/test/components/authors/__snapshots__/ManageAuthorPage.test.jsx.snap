// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ManageAuthorPage /> component render component without crashing 1`] = `
<Fragment>
  <Prompt
    message={[Function]}
    when={false}
  />
  <div
    className="top-container"
  >
    <h1>
      Manage Author
    </h1>
  </div>
  <AuthorForm
    allAuthors={
      Array [
        Object {
          "firstName": "Cory",
          "id": "cory-house",
          "lastName": "House",
        },
        Object {
          "firstName": "Scott",
          "id": "scott-allen",
          "lastName": "Allen",
        },
      ]
    }
    author={
      Object {
        "firstName": "",
        "id": "",
        "lastName": "",
      }
    }
    errors={Object {}}
    handleOnChange={[Function]}
    handleOnFocus={[Function]}
    handleOnSave={[Function]}
    isSaving={false}
  />
</Fragment>
`;
